rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,394 @@
+
+### TensorFlow API modularity
+
+There are other side-benefits if we split the repository. Currently, Keras
+has to rely on a number of private TensorFlow APIs. However, a litmus test
+of the quality of the public TensorFlow low-level APIs is that they should
+be strictly sufficient to a higher-level API like Keras.
+After splitting the repository, Keras will have to import TensorFlow and 
+rely exclusively on public APIs. If Keras still ends up using TensorFlow
+private features, it  might be an indication of tight coupling of
+implementation details. If certain private features are extensively used,
+we might want to consider exposing them  as public low level API.
+
+This design is also aligned with the design for
+[Modular TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md), 
+which splits the TensorFlow project into smaller components that are not
+tightly coupled together.
+
+
+## Design Proposal
+
+### New location of the code
+
+GitHub: the code will live at [keras-team/keras](https://github.com/keras-team/keras), 
+joining the other Keras SIG projects and replacing the current external Keras 
+codebase. `tf.Keras` will also replace Keras on PyPI.
+
+Also considered: `tensorflow/keras`.
+
+Pros:
+1. Under the umbrella of Keras SIG, which hosts all other Keras related projects
+like keras-application, KerasTuner etc.
+1. Lots of existing followers on keras-team, who may not be easily migrated to 
+TF project.
+1. Can't easily delete keras project, which already have tons of stars and 
+incoming reference links. Continued existence of external Keras code will create
+confusion ("why is there tensorflow/keras AND keras-team/keras?").
+
+Cons:
+1. The repo isn't under the same organization as tensorflow, which makes it hard
+to manage issues/PRs and references across the organization.
+1. Existing issue/PR under the same org can be transferred easily, but not cross the different org. See [here](https://help.github.com/en/github/managing-your-work-on-github/transferring-an-issue-to-another-repository).
+
+### Source of Truth
+
+TensorFlow uses a Google-internal code repository as its source of truth. Every PR
+submitted though GitHub is converted to a Google-internal change first,
+submitted through the internal system, and then copied to GitHub as commits.
+At the same time, PR is marked as merged with the corresponding commit hash.
+
+Likewise, issue tracking and code review takes place through Google-internal tools.
+
+For Keras, since we are trying to promote community engagement, we hope to use 
+GitHub as source of truth. This will have the following implications:
+
+* We expect the majority of the code development/contribution from GitHub
+and the dev tools / tests / scripts should focus on the GitHub development use
+case. See below for more details.
+* Keras CI/presubmit build for the GitHub repo should target a stable PIP 
```

Comment:
This has the disadvantage of slowing down keras's adoption of new TF features and APIs.

This document should clarify when would CI against tf stable be expected to break, etc.

Reply:
(also important to highlight the release process for TF)

Reply:
It's not supposed to slow down adoption of new features compared to targeting an unpinned nightly all the time. Maybe it's not clear in the document, but if a pull request requires a new feature in tensorflow nightly, the person making the pull request can change the targeted version of tensorflow in the CI. So we don't have to wait at all.

Do I get it right or you were referring to another problem?